### PR TITLE
Error text is shown just above the answer options in boolean, single and multiple choice view.

### LINKED ISSUE #1586: Move error message for boolean/single/multiple choice type widgets to above the choices
**Describe the bug**
When error message get shown below the sdc components like boolean choice, single choice etc, there is not enough margin in between view and error text. 

**Component**
SDC library

**Expected behavior**
It is expected to show enough margin in between view and error texts.

**Screenshots**
Screen Shot 2022-09-08 at 11 46 56 AM

**Smartphone (please complete the following information):**
 - Device: [e.g. Pixel4a emulator]
 - Android version: [e.g. Settings -> About phone -> Android version]
 - Build number: [e.g. Settings -> About phone -> Build number]

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1586

**Description**
Single choice, Boolean choice, Multiple choice error text was not having top margin.
Top margin is given to error text.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by santosh-pingle** [comment]
@shelaghm 
Before change 
Screen Shot 2022-10-07 at 12 40 48 PM
Screen Shot 2022-10-07 at 12 40 35 PM
Screen Shot 2022-10-07 at 12 40 24 PM


After change
Screen Shot 2022-10-07 at 12 29 29 PM
Screen Shot 2022-10-07 at 12 28 59 PM
Screen Shot 2022-10-07 at 12 28 29 PM

**Comment by shelaghm** [comment]
@santosh-pingle @jingtang10 The top margin padding looks good. 

However, I'm a bit concerned that the error message for Single choice, Boolean choice, Multiple choice error text is being displayed at the bottom of the list. I was assuming the error text for these components would be displayed in the `instructions`  text field. What are your thoughts on this?

Example:
Screen Shot 2022-10-07 at 11 36 51

**Comment by santosh-pingle** [comment]
> @santosh-pingle @jingtang10 EDIT**: Please reference padding suggestions in #1586
> 
> > Spacing: @santosh-pingle what is the line spacing on the text element for the EntryFormat / error text?
> > According to our M3 guidelines, the line height of the text is 16px, the font is 12px and with 4px padding. I'm guessing the issue is that the text line spacing is not 16px, which is why it looks like less padding than it should be?
> 
> However, I'm a bit concerned that the error message for Single choice, Boolean choice, Multiple choice error text is being displayed at the bottom of the list. I was assuming the error text for these components would be displayed in the `instructions` text field. What are your thoughts on this?
> 
> Example: Screen Shot 2022-10-07 at 11 36 51

@shelaghm 
Actually, there are other views where we shows error text at the bottom.

**Comment by jingtang10** [comment]
> > @santosh-pingle @jingtang10 EDIT**: Please reference padding suggestions in #1586
> > > Spacing: @santosh-pingle what is the line spacing on the text element for the EntryFormat / error text?
> > > According to our M3 guidelines, the line height of the text is 16px, the font is 12px and with 4px padding. I'm guessing the issue is that the text line spacing is not 16px, which is why it looks like less padding than it should be?
> > 
> > 
> > However, I'm a bit concerned that the error message for Single choice, Boolean choice, Multiple choice error text is being displayed at the bottom of the list. I was assuming the error text for these components would be displayed in the `instructions` text field. What are your thoughts on this?
> > Example: Screen Shot 2022-10-07 at 11 36 51
> 
> @shelaghm Actually, there are other views where we shows error text at the bottom.

@santosh-pingle what do you mean?

**Comment by jingtang10** [comment]
just to clarify what santosh meant (just discussing with him)

We have 2 types of widgets

1. widgets that are derivatives of the text input layout from material. these include all the text input widgets (integer, single line, multi line etc), the date and date time inputs (because we use the text field to display the date and time), drop down (again, we use a text field to display the selected value). because we use the text input layout from material for these widgets, they already have the error text correctly laid out 4dp below the text input.
2. widgets that have custom layout that we have written - single, boolean, multiple choices and potentially others.

just a note - we don't necessarily have to use the material error text either in the first type of widgets, we could also create another error text field and place it elsewhere instead of using the text input layout's error field. but that's a different question.

**Comment by shelaghm** [comment]
@santosh-pingle @jingtang10 I think the 4dp padding looks fine on the radio button. It is consistent with M3 guidelines, so let's keep this as Santosh has implemented and not make further adjustments on the padding.

**Comment by shelaghm** [comment]
> 2\. widgets that have custom layout that we have written - single, boolean, multiple choices and potentially others.

Thanks for the context, this is very helpful.

**Comment by jingtang10** [comment]
any update here @santosh-pingle?

**Comment by santosh-pingle** [comment]
> any update here @santosh-pingle?

These changes are obsolete now. 
As per discussion which we had in ux design call, error text will be shown just below the instructions.
I have not started working on it yet, as there were other priority task.

Can you please help me to know the priority of this task vs sdc module testing which I supposed to start on tomorrow after ux testing build?

**Comment by santosh-pingle** [comment]
@shelaghm 
Screen Shot 2022-10-31 at 4 14 25 PM
Screen Shot 2022-10-31 at 4 14 50 PM
Screen Shot 2022-10-31 at 4 15 10 PM

File: datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderViewTest.kt [hunk 1]
```diff
@@ -272,6 +272,22 @@ class QuestionnaireItemHeaderViewTest {
       .isEqualTo(view.context.applicationContext.resources.getString(R.string.space_asterisk))
   }
 
+  @Test
+  fun `shows custom error text`() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderViewTest.kt:276]: not sure what is meant by custom here. can you please remove? same for the next test case.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
can you update the PR name please? thanks

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderView.kt [hunk 2]
```diff
@@ -62,6 +63,22 @@ internal class QuestionnaireItemHeaderView(context: Context, attrs: AttributeSet
     // questionnaire.
     visibility = getViewGroupVisibility(prefix, question, hint)
   }
+
+  fun showCustomErrorText(errorText: String? = null, isErrorTextVisible: Boolean = true) {
+    fun TextView.showError() {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderView.kt:68]: working on this change.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderView.kt [hunk 3]
```diff
@@ -62,6 +63,22 @@ internal class QuestionnaireItemHeaderView(context: Context, attrs: AttributeSet
     // questionnaire.
     visibility = getViewGroupVisibility(prefix, question, hint)
   }
+
+  fun showCustomErrorText(errorText: String? = null, isErrorTextVisible: Boolean = true) {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderView.kt:67]: working on this change.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderViewTest.kt (hunk 1 above)
Comment (santosh-pingle) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderViewTest.kt:276]: working on this change.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderView.kt [hunk 4]
```diff
@@ -62,6 +63,19 @@ internal class QuestionnaireItemHeaderView(context: Context, attrs: AttributeSet
     // questionnaire.
     visibility = getViewGroupVisibility(prefix, question, hint)
   }
+
+  fun showErrorText(errorText: String? = null, isErrorTextVisible: Boolean = true) {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemHeaderView.kt:67]: add documentation here to say that this is showing error in the header. and widgets could either use this, or use another view (i.e. TextInputLayout's error field) to display error.

**Review by jingtang10 — APPROVED** [review]

**Comment by santosh-pingle** [comment]
Closing this pr as cla/google check is failing, which ask other author approval which is not required.
Instead created PR#1700

**Pull request closed without merging**